Menu Manager base system

JodyLS27 opened the pull request:
# Info
This is a base system for the menus in the game so that one menu at a time is loaded into memory.
There are two base classes `i_switch_menu` and `i_menu`. There are also, two menu classes for now, `main_menu` and `settings_menu` and the `menu_manager` is used to manage which menu is currently loaded by controlling the switching and the calling from the `game_manager`.

JodyLS27 committed:
Adding menu base class
- Adding Base class functions.

JodyLS27 committed:
Menu Manager
- Adding TODOs for later
- Adding some Functions to the menu_manager
- Added current state of type MenuBase

JodyLS27 committed:
Adding main Menu
- Added Main Menu Header and Source files
- Added TODO Tasks
- Cleaned up Includes

JodyLS27 committed:
Updating the interface
- Renamed the interface to i_menu
- Moved the interface into /include/menu

JodyLS27 committed:
State Machine Test Statments
- Added some console logs to see how the pointer is moving around
- Adjusted the default value of the unique pointer in the menu_manager
- set the loading menu to main_menu

JodyLS27 committed:
Clean up
- Renamed gameManager => game_manager
- Cleaned up uneeded code in game_manager
- Adjusted main.cpp to the new changes

JodyLS27 committed:
Solving build errors
- Had to include the source files in teh CMakeLists.txt under /src
- Adjusted some code while debugging and solving issues

JodyLS27 committed:
Testing logic
- Testing to see what the current menu is and when its being destroyed

JodyLS27 committed:
Basic Test logic
- Testing to see if we can switch menus
- Adding Virtual Destructore for the menu interface
- Added logging for testing
- Added Destructor overloads for i_menu

JodyLS27 committed:
Logic update
- Adding a switch state to change menus
- Adding logic to change states in the main_menu update() function
- Adding test code sections for logic testing

JodyLS27 committed:
Fixed merge conflicts

JodyLS27 committed:
Menu Switching setup
- Updated switch_menu to recive a pointer of Menu base class
- Updated switch_menu to include a unique_ptr reset
- Updated `current_menu` `Update()` to take a class object
- Updated the `Menu` base class to take a `MenuManager` class reference.
- Updated Logic in `MainMenu` `update()` to switch menus based on inputs.

- Added a getter and setter for the `current_option` and `next_option` parameters in `menu_manager`
- Added `m_current_option` to the `MenuManager` class

- Removed `current_option` from the `MainMenu` class

JodyLS27 committed:
Menu Base Update
- Updated `i_menu` base class to take in a pointer of tpe `MenuManager`
- Updated the `MainMenu` class, `update()` function to tke in a pointer and not a reference to a class.

JodyLS27 committed:
Adding settings
- Added settings_menu header and source file

- Updated the logic check in MainMenu::update()

- Removed this in MainMenu::update(this) for the actual class

JodyLS27 committed:
Updated switch_menu function
- Updated MenuManager::switch_menu() to take in a unique_ptr instead of a Raw pointer
  - The Unique_ptr is of type Mene which is the base class

JodyLS27 committed:
Updating Virtual class abstraction
- Added forward declaration of `MenuManager` in  `i_menu.hpp`

- Updated `menu::update()` to take a referene to `menu::MenuManager`
- Updated `settings_menu` and `main_menu` to the `menu::MenuManager` reference

JodyLS27 committed:
Code error cleanup
- Added Forward declare of `Menu` in `menu_manager.hpp`
- Added Forward declare of `MenuManager` in `i_menu.hpp`
- Added some test `cout` to test values in `main_menu` and `settings_menu`

- Updated Unique_ptr with a removed namespace call in `menu_manager.hpp`
- Updated `menu::MenuManager::update()` to pass an object reference to the current menu `update()`

- Removed the virtual destructor from `i_menu.hpp`, (No cleanup is needed)
- Removed ~destructor override in `main_menu.hpp` and `settings_menu.hpp`

JodyLS27 committed:
Slight refactor
- Added a default value to the m_current_option in menu_manager.cpp
- Cleaned up some std::cout messages in settings
- Updated the order in the game loop in game_manager

JodyLS27 committed:
updating menu class naming
- Added I in front of Menu to show its an Interface
- Updated classes that are using this interface

JodyLS27 committed:
Add Diagrams
- Added class diagram for MenuManager
- updated folder naming structure

JodyLS27 committed:
Implement default destructor
- Abstract class Virtual destructor

JodyLS27 committed:
Added new ISwitchMenu
- Added new header for safe menu switching from menu classes

JodyLS27 committed:
ISwitchMenu Logic
- Added Inherit of ISwitchMenu
- Added next_menu ptr

- Removed `m_next_option` and `m_current_option`
- Remvoed `get` and `set` functions for above variables

JodyLS27 committed:
Bringing back Menu Options
- Brought back a menu selection screen
- Setting current and next menu variables in menu_manager

JodyLS27 committed:
Adjusting game draw
- Added `draw()` to `MenuManager` so I am able to call the current menus draw.
- Added the menu text to `draw()` in `main_menu.cpp`

- Updated `game_manager.cpp` to call `MenuManager::draw()`

- Removed some uneeded code from `MainMenu::update()`

JodyLS27 committed:
Updating Settings draw
- Added basic menu setup to the draw() function.

JodyLS27 committed:
Cleanup
- Added some new `TODOs` for next implementations

- Removed Uneeded `std::cout` in menus
- Removed Uneeded `TODOs` and comments.

Kouros26 commented on include/core/menu_manager.hpp:
```diff
@@ -0,0 +1,46 @@
+	// forward declare
+	class IMenu;
+
+	class MenuManager : public menu::ISwitchMenu
```
you don't need to write ```menu::ISwitchMenu``` since you're in the ```menu``` namespace.

Kouros26 commented on src/core/game_manager.cpp:
```diff
@@ -0,0 +1,31 @@
+
+void game_manager::update(menu::MenuManager& menu_manager)
+{
+	menu_manager.init();
```
you should call this in the ```game_manager::init()``` function

JodyLS27 replied:
✅ Implemented.

Kouros26 commented on src/menu/main_menu.cpp:
```diff
@@ -0,0 +1,40 @@
+	if (menu_manager.get_current_option() != menu_manager.get_next_option())
+	{
+		if (menu_manager.get_next_option() == 2)
+			menu_manager.switch_menu(std::make_unique<menu::SettingsMenu>());
```
this logic should be in the menu manager not in one of the menus

JodyLS27 replied:
Thanks @Kouros26, Should this rather be in the `menu_manager::update()` function and if we are not switching, we are running the current Menus `update()` ?

Kouros26 replied:
yep exactly, but no need to check if we are switching or not, run the current menu's update function regardless

JodyLS27 replied:
I am going to look into this.
From my understanding of a state machine, we want this in each menu class as each one can switch to different classes based on the logic Constraint.

_(i.e: `Main Menu` can switch to `Settings`, `Game` and `Top Score` But Settings can only go back to `Main Menu` and cannot go to `Game`)_
So placing this logic in each State (`Sub Class`) should be better ?

Kouros26 commented on src/menu/settings_menu.cpp:
```diff
@@ -0,0 +1,40 @@
+	if (menu_manager.get_current_option() != menu_manager.get_next_option())
+	{
+		if (menu_manager.get_next_option() == 1)
+			menu_manager.switch_menu(std::make_unique<menu::MainMenu>());
```
same here

Kouros26 commented:
Also forgot to add that your ```IMenu``` class shouldn't know about the ```MenuManager``` class

JodyLS27 committed:
Game Manger Init
- Updated game_manager::init() to run the MenuManager::init() function.

JodyLS27 committed:
cleanup
- Adding const to getters
- Removing forward declare in i_menu
- Removing MenuManger reference from the Update() function in menus.

JodyLS27 committed:
State Machine cleanup
- Adding `set_menu_controller` Function
- Added a Reference to the `MenuManager` class

- Update: Todo Cleanup and Comments.

JodyLS27 commented:
> Also forgot to add that your `IMenu` class shouldn't know about the `MenuManager` class

yes, Will Implement this now, Thanks man.

JodyLS27 committed:
Function Docs
- Added Function docs for game_manager::init() and game_manager::update()

JodyLS27 committed:
Adding Function docs
- Adding Function docs for Clarity going forward
- Added Docs to menu_manager.hpp.

JodyLS27 committed:
Update MenuManager to MenuController

JodyLS27 committed:
Renamingg i_switch_menu to i_menu_switcher

JodyLS27 committed:
Updating Menu Controller
- Renaming current_menu to be m_current_menu
- Renaming
ext_menu to be m_next_menu

- Updated i_menu_switcher to with function docs and TODO Notes
- Added constructor to IMenu

JodyLS27 committed:
Updating Menu class Diagram
- Cleanup Diagram and updating flow

JodyLS27 committed:
Updating game manager Main function
- Renamed update() to 
un() to show this is where the main program starts

JodyLS27 committed:
Updating CMakeLists.txt Remove include
- Removed the include to the `FetchSFML.cmake` line

JodyLS27 committed:
Updating game_manager - Cleanup
- Updated `start()` to `update()`
- changed `run()` to `start()` in the source file

JodyLS27 committed:
- Removed Uneeded function for now
- Updated main to run the `gameManager::start()` and `game_manager::exit()` functions

JodyLS27 committed:
Updating CMake files
- Updated the version of the Project to be `0.1.0` as its a better starting point with semantic versioning::`/CMakeLists.txt`
- Updated `/cmake/FetchSFML.cmake` to allow SFML to be used locally after being cloned down and not Reference GitHub during each built.

JodyLS27 committed:
Updating main
- Cleanup of imports and removed the include of `main.h` and it's not in the project
- Clight cleanup of code padding
